src/main/kotlin/com/cultureamp/eventsourcing/BlockingAsyncEventProcessorWaiter.kt

```diff
@@ -0,0 +1,47 @@
+package com.cultureamp.eventsourcing
+
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withTimeout
+
+class BlockingAsyncEventProcessorWaiter<M : EventMetadata>(
+    private val eventProcessors: List<BookmarkedEventProcessor<M>>,
+    private val maxWaitMs: Long,
+    private val pollWaitMs: Long,
+    private val logger: (String) -> Unit = System.out::println,
+) {
+    fun waitUntilProcessed(events: List<SequencedEvent<M>>) {
+        val projectorToMaxRelevantSequence: Map<BookmarkedEventProcessor<M>, Long> = eventProcessors.associateWith { eventProcessor ->
+            val eventProcessorEventTypes = eventProcessor.sequencedEventProcessor.domainEventClasses().toSet()
+            val relevantSequencedEvents = events.filter { eventProcessorEventTypes.contains(it.event.domainEvent::class) }
+            val maxRelevantSequence = relevantSequencedEvents.lastOrNull()?.let { it.sequence }
+            maxRelevantSequence
+        }.filterNotNullValues()
```

Comment:
just pointing out that this code is pretty dense and hard to read. no specific suggestion though

Reply:
Channelling my old Scala days on all of this functional transformation flimflams